File: 1639-number-of-ways-to-form-a-target-string-given-a-dictionary/1639-number-of-ways-to-form-a-target-string-given-a-dictionary.py
class Solution:
    def numWays(self, words: List[str], target: str) -> int:
        n, t, mod = len(words[0]), len(target), 10**9 + 7
        memo = [1] + [0] * t 
        for i in range(n):
            count = Counter(word[i] for word in words)
            for j in range(min(i + 1, t), 0, -1):
                memo[j] = (memo[j] + count[target[j - 1]] * memo[j - 1]) % mod
        return memo[-1]
      

File: 1639-number-of-ways-to-form-a-target-string-given-a-dictionary/1639-number-of-ways-to-form-a-target-string-given-a-dictionary.js
/**
 * @param {string[]} words
 * @param {string} target
 * @return {number}
 */
var numWays = function(words, target) {
  const [m, n, t, mod] = [words.length, words[0].length, target.length, 1e9 + 7];
  const a = 'a'.charCodeAt();  
  const dp = new Array(t + 1).fill(0);
  dp[0] = 1;
  
  for (let i = 0; i < n; i++) {
    const count = new Array(26).fill(0);
    for (let word of words)
      count[word.charCodeAt(i) - a]++;
    
    for (let j = Math.min(i + 1, t); j > 0; j--) 
      dp[j] = (dp[j] + count[target.charCodeAt(j - 1) - a] * dp[j - 1]) % mod;
  }
  
  return dp[t];
};